contracts/buyAndBurn/BuyAndBurnSwapImplementation.sol

```diff
@@ -0,0 +1,402 @@
+        ISwapHelper.Quote memory quote = swapHelper.quoteSlippageExactInputSingle(
+            ISwapHelper.QuoteSlippageExactInputSingleParams({
+                tokenIn: tokenIn,
+                tokenOut: tokenOut,
+                fee: params.fee,
+                amountIn: params.amountIn,  // Quote the FULL amount (greedy)
+                sqrtPriceX96: currentSqrtPriceX96,
+                liquidity: currentLiquidity,
+                maximumSlippagePercentage: params.singleBatchImpactThreshold
+            })
+        );
+
+        // The quote returns the maximum amount we can swap within price impact limits
+        amountSwapIn = quote.amountToPay;  // This will be ≤ params.amountIn
+
+        // CRITICAL BRANCHING: Check if we can swap the ENTIRE amount
+        if (amountSwapIn == params.amountIn) {
+            // YES PATH: No unused TokenIn - entire amount can be swapped within impact threshold
+            // This is the pure greedy case: maximum burn, no LP needed
+
+            // Approve swapHelper if tokenIn is ERC20
+            if (params.tokenIn != VANA) {
+                IERC20(params.tokenIn).forceApprove(address(swapHelper), amountSwapIn);
+            }
+
+            // Execute swap with singleBatchImpactThreshold (same as quote)
+            (uint256 amountSwapInUsed, uint256 amountReceived) = swapHelper.slippageExactInputSingle{
+                    value: params.tokenIn == VANA ? amountSwapIn : 0
+                }(
+                ISwapHelper.SlippageSwapParams({
+                    tokenIn: params.tokenIn,    // Use original address (VANA or ERC20)
+                    tokenOut: params.tokenOut,  // Use original address (VANA or ERC20)
+                    fee: params.fee,
+                    recipient: address(this),
+                    amountIn: amountSwapIn,
+                    maximumSlippagePercentage: params.singleBatchImpactThreshold  // ✅ Use same threshold
+                })
+            );
+
+            amountSwapIn = amountSwapInUsed;
+            amountSwapOut = amountReceived;
+
+            // Handle WVANA wrapping for tokenOut based on what SwapHelper returned
+            if (params.tokenOut == VANA && amountSwapOut > 0) {
+                // Check if swap returned WVANA or native VANA
+                uint256 wvanaBalanceAfter = IERC20(address(WVANA)).balanceOf(address(this));
+                uint256 wvanaReceived = wvanaBalanceAfter > wvanaBalanceBefore ? wvanaBalanceAfter - wvanaBalanceBefore : 0;
+
+                // If we received less WVANA than expected, wrap the native VANA we received
+                if (wvanaReceived < amountSwapOut) {
+                    uint256 ethToWrap = amountSwapOut - wvanaReceived;
+                    require(address(this).balance >= ethToWrap, "Insufficient ETH received from swap");
+                    WVANA.deposit{value: ethToWrap}();
+                }
+            }
```

Comment:
This logic is not needed. If `params.tokenOut == VANA`, `slippageExactInputSingle` will returns VANA to this contract.